packages/core/src/component.test.ts

```diff
@@ -0,0 +1,311 @@
+                    return this.state;
+                }
+            }
+            const component = new TestComponent()
+
+            component.callUpdate()
+            expect(component.getState()).toBe(undefined)
+
+            const newState = {a: 1}
+            component.callUpdate(newState)
+            expect(component.getState()).toBe(newState)
+
+            expect(component.callRender).toHaveBeenCalledTimes(2)
+        })
+        it("pass correct params to the subscribeWithSelector", () => {
+            resetDataLayer();
+
+            const selector = identity
+            const store = createEmptyStore(null)
+
+            class TestComponent extends Component {
+                selector = selector
+                store = store
+                getUpdate() {
+                    return this.update
+                }
+                getStatesAreEqual() {
+                    return this.statesAreEqual
+                }
+            }
+            const component = new TestComponent()
+            component.connectedCallback()
+
+            expect(subscribeWithSelector).toHaveBeenCalledWith(
+                component.store,
+                component.selector,
+                component.getUpdate(),
+                component.getStatesAreEqual()
+            )
+        })
+    })
+    describe("render", () => {
+        it("calls beforeRender and afterRender in callRender", () => {
+            class TestComponent extends Component {
+                beforeRender = jest.fn()
+                afterRender = jest.fn()
+
+                callRender() {
+                    super.callRender()
+                }
+            }
+            const component = new TestComponent()
+            component.callRender()
+
+            expect(component.beforeRender).toHaveBeenCalledTimes(1)
+            expect(component.afterRender).toHaveBeenCalledTimes(1)
+        })
+        it("calls render with root or shadowRoot depends on the isShadow prop", () => {
+            const shadowRoot = {}
+            const root = {}
+
+            class TestComponent extends Component {
+                render = jest.fn()
+                shadowRoot = shadowRoot as any
+                root = root as any
+
+                callRender() {
+                    super.callRender()
+                }
+            }
+            class ShadowComponent extends TestComponent {
+                isShadow = true
+            }
+            const shadowComponent = new ShadowComponent()
+            shadowComponent.callRender()
+
+            expect(shadowComponent.render).toHaveBeenCalledWith(shadowRoot)
+
+            class NonShadowComponent extends TestComponent {
+                isShadow = false
+            }
+            const component = new NonShadowComponent()
+            component.callRender()
+
+            expect(component.render).toHaveBeenCalledWith(root)
+        })
+    })
+
+    it("throws error in default getName", () => {
```

Comment:
Could be split by the several files?

Reply:
split what?

Reply:
`component.test.ts`

Reply:
why? all tests are related to the Component. i don't think it's a good idea

Reply:
still - all scenarios are related to the single file